### PR TITLE
refactor hash load balancer route generation

### PR DESCRIPTION
**Pull request by ramaraochavali**
This PR
- Refactors consistent hash route generation logic to reuse code
- Removes unnecessary loops and allocations.

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 1]
```diff
@@ -68,6 +68,8 @@ const prefixMatchRegex = `((\/).*)?`
 
 var notimeout = durationpb.New(0)
 
+type DestinationHashMap map[*networking.HTTPRouteDestination]*networking.LoadBalancerSettings_ConsistentHashLB
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:71]: Using pointer type key is very dangerous

**Review by hzxuzhonghu — COMMENTED** [review]
@ramaraochavali I need to take a close look, since it is refactoring critical code.

And if  it is  a perf improve, better provide a benchmark

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:71]: This is existing one only. I just introduced struct.

**Comment by ramaraochavali** [comment]
> And if it is a perf improve, better provide a benchmark

I am not expecting a significant perf improvement

>  I need to take a close look, since it is refactoring critical code.

Sure take your time

**Review by hzxuzhonghu — COMMENTED** [review]
@ramaraochavali Can you mark some critical improve points,  you know it is super painful to review a refactoring patch

**Comment by ramaraochavali** [comment]
Let me split this in to multiple PRs so that it is easier to review.

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 2]
```diff
@@ -93,45 +93,42 @@ type VirtualHostWrapper struct {
...
 			delete(serviceRegistry, service.Hostname)
 		}
 	}
 
-	hashByService := map[host.Name]map[int]*networking.LoadBalancerSettings_ConsistentHashLB{}
 	for _, svc := range serviceRegistry {
 		for _, port := range svc.Ports {
 			if port.Protocol.IsHTTP() || util.IsProtocolSniffingEnabledForPort(port) {
-				hash, destinationRule := getHashForService(node, push, svc, port)
+				hash, destinationRule := hashForService(node, push, svc, port)
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:126]: @hzxuzhonghu I know you preferred previous name. But changed to be consistent with `hashForVirtualService`

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 3]
```diff
@@ -250,35 +245,25 @@ func buildSidecarVirtualHostsForVirtualService(
 	return out
 }
 
-func buildSidecarVirtualHostsForService(
-	serviceRegistry map[host.Name]*model.Service,
-	hashByService map[host.Name]map[int]*networking.LoadBalancerSettings_ConsistentHashLB,
+func buildSidecarVirtualHostForService(svc *model.Service,
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:248]: @hzxuzhonghu This is the main change left over from previous PR i.e. `buildSidecarVirtualHostForService` for single service as its name suggests and move inside the loop above

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 4]
```diff
@@ -1275,21 +1260,12 @@ func getHashForService(node *model.Proxy, push *model.PushContext,
...
 
-func hashForVirtualService(push *model.PushContext, node *model.Proxy, virtualService config.Config,
-	serviceRegistry map[host.Name]*model.Service,
-) (DestinationHashMap, []*model.ConsolidatedDestRule) {
+func hashForVirtualService(push *model.PushContext, node *model.Proxy, virtualService config.Config) (DestinationHashMap, []*model.ConsolidatedDestRule) {
 	hashByDestination := DestinationHashMap{}
 	destinationRules := make([]*model.ConsolidatedDestRule, 0)
 	for _, httpRoute := range virtualService.Spec.(*networking.VirtualService).Http {
 		for _, destination := range httpRoute.Route {
-			hostName := destination.Destination.Host
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:1285]: configNamespace is not used so removed it.

**Review by ramaraochavali — COMMENTED** [review]
@hzxuzhonghu Thank you for the previous commit review. I rebased this on that commit and highlighted the changes. PTAL when you get chance - not urgent.

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 5]
```diff
@@ -1248,8 +1233,8 @@ func consistentHashToHashPolicy(consistentHash *networking.LoadBalancerSettings_
 	return nil
 }
 
-func getHashForService(node *model.Proxy, push *model.PushContext,
-	svc *model.Service, port *model.Port,
+func hashForService(node *model.Proxy, push *model.PushContext, svc *model.Service,
+	port *model.Port,
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:1237]: Nit: for multi paramas, i prefer one per line, which look clear for me

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:248]: LGTM

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 6]
```diff
@@ -1248,8 +1233,8 @@ func consistentHashToHashPolicy(consistentHash *networking.LoadBalancerSettings_
 	return nil
 }
 
-func getHashForService(node *model.Proxy, push *model.PushContext,
-	svc *model.Service, port *model.Port,
+func hashForService(node *model.Proxy, push *model.PushContext, svc *model.Service,
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:1236]: nit: keep consistent of the order with other functions 1. push context, 2. proxy, 3 others (svc, port)

**Review by hzxuzhonghu — COMMENTED** [review]
Good improvement

**Comment by ramaraochavali** [comment]
@hzxuzhonghu addressed comments. PTAL

**Review by hzxuzhonghu — APPROVED** [review]
Thanks

**Pull request merged**